File: 34 Find First and Last Position of Element in Sorted Array/34 Find First and Last Position of Element in Sorted Array.go
func searchRange(nums []int, target int) []int {
    start := find(nums, target, true)
    if start == -1 {
        return []int{-1, -1}
    }
    end := find(nums, target, false)
    return []int{start, end}
}

func find(nums []int, target int, start bool) int {
    n := len(nums)
    left, right := 0, n-1
    res := -1
    for left <= right {
        mid := left + (right - left) / 2
        if nums[mid] < target {
            left = mid+1
        } else if nums[mid] > target {
            right = mid-1
        } else {
            res = mid
            if start {
                right = mid-1
            } else {
                left = mid+1
            }
        }
    }
    return res
}

File: 34 Find First and Last Position of Element in Sorted Array/34 Find First and Last Position of Element in Sorted Array.py
"""
Given an array of integers nums sorted in ascending order, find the starting and ending position of a given target value.
If target is not found in the array, return [-1, -1].
You must write an algorithm with O(log n) runtime complexity.

Example 1:

Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]

Example 2:

Input: nums = [5,7,7,8,8,10], target = 6
Output: [-1,-1]

Example 3:

Input: nums = [], target = 0
Output: [-1,-1]
"""
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return [-1, -1]
        left = 0
        right = len(nums) - 1
        while left < right:
            mid = - (- (left + right) // 2)
            if target >= nums[mid]:
                left = mid
            else:
                right = mid - 1
        if nums[left] == target:
            upper = left
        else:
            return [-1, -1]
        left = 0
        right = len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            if target <= nums[mid]:
                right = mid
            else:
                left = mid + 1
        lower = left
        return [lower, upper]
